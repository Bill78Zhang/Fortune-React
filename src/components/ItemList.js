import React from 'react';
import IosCafeOutline from 'react-ionicons/lib/IosCafeOutline';
import IosAdd from 'react-ionicons/lib/IosAdd';
import IosClose from 'react-ionicons/lib/IosClose';
import PropTypes from 'prop-types';

export default function ItemList({items, onModifyItem, onDeleteItem}) {
  return (
    <div>
      <ul className='list-group list-group-flush'>
        {items.map(item => {
          return (
            <li
              className='list-group-item  d-flex justify-content-between align-item-center'
              key={item.id}>
              <span className='col-1'>
                <IosCafeOutline
                  className='rounded-circle'
                  font-size='100px'
                  style={{background: '#007bff', padding: '2px'}}
                  color={'#fff'}
                />
              </span>
              <span className='col-5'>{item.title}</span>
              <span className='col-2 font-weight-bold'>
                {item.category.type === 'outcome'
                  ? `-${item.price}`
                  : item.price}
                $
              </span>
              <span className='col-2'>{item.date}</span>
              <IosAdd
                className='rounded-cycle'
                font-size='100px'
                style={{background: '#28a745', padding: '2px'}}
                color={'#fff'}
                onClick={onModifyItem}
              />
              <IosClose
                className='rounded'
                font-size='100px'
                style={{background: '#dc3545', padding: '2px'}}
                color={'#fff'}
                onClick={() => onDeleteItem(item)}
              />
            </li>
          );
        })}
      </ul>
    </div>
  );
}

ItemList.propTypes = {
  items: PropTypes.array.isRequired,
  onModifyItem: PropTypes.func.isRequired,
  onDeleteItem: PropTypes.func.isRequired
};

ItemList.defaultProps = {
  onModifyItem: () => {}
};
