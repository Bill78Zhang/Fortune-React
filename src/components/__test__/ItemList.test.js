import {shallow} from 'enzyme';
import ItemList from '../ItemList';
import React from 'react';
import {exportAllDeclaration} from '@babel/types';
const props = {
  items: [
    {
      id: 1,
      title: 'Buy Milk',
      price: 40,
      date: '2020-5-20',
      category: {
        id: 1,
        type: 'outcome',
        name: 'shopping',
        iconName: 'IosCafeOutline'
      }
    },
    {
      id: 2,
      title: 'Electrity',
      price: 130,
      date: '2020-5-1',
      category: {
        id: 2,
        type: 'income',
        name: 'Gas',
        iconName: 'IosCafeOutline'
      }
    },
    {
      id: 3,
      title: 'Electrity',
      price: 130,
      date: '2020-5-1',
      category: {
        id: 3,
        type: 'income',
        name: 'Gas',
        iconName: 'IosCafeOutline'
      }
    },
    {
      id: 4,
      title: 'Electrity',
      price: 130,
      date: '2018-4-1',
      category: {
        id: 4,
        type: 'income',
        name: 'Gas',
        iconName: 'IosCafeOutline'
      }
    }
  ],
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn()
};

let wrapper;
describe('test ItemList Component', () => {
  beforeEach(() => {
    wrapper = shallow(<ItemList {...props} />);
  });
  it('should render the component to match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('render correct items length', () => {
    expect(wrapper.find('.list-group-item').length).toEqual(props.items.length);
  });
  it('shoule trigger the correct function callback', () => {
    const firstItem = wrapper.find('.list-group-item').first();
    firstItem.find('.rounded-cycle').simulate('click');
    expect(props.onModifyItem).toHaveBeenCalled();
    firstItem.find('.rounded').simulate('click');
    expect(props.onDeleteItem).toHaveBeenCalledWith(props.items[0]);
  });
});
