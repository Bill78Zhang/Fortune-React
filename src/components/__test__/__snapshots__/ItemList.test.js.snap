// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ItemList Component should render the component to match the snapshot 1`] = `ShallowWrapper {}`;
